security/advancedtls/examples/credential_reloading_from_files/client/main.go

```diff
@@ -92,34 +91,20 @@ func main() {
 
 	// Send the requests every 0.5s. The credential is expected to be changed in
 	// the bash script.
-	ticker := time.NewTicker(500 * time.Millisecond)
-	done := make(chan bool)
-	go func() {
-		for {
-			select {
-			case <-done:
-				return
-			case <-ticker.C:
-				// Make a connection using the credentials.
-				ctx, cancel := context.WithTimeout(context.Background(), defaultConnTimeout)
-				conn, err := grpc.DialContext(ctx, address, grpc.WithTransportCredentials(clientTLSCreds))
-				if err != nil {
-					log.Fatalf("grpc.DialContext to %s failed: %v", address, err)
-				}
-				client := pb.NewGreeterClient(conn)
-				_, err = client.SayHello(ctx, &pb.HelloRequest{Name: "gRPC"}, grpc.WaitForReady(true))
-				if err != nil {
-					log.Fatalf("client.SayHello failed: %v", err)
-				}
-				cancel()
-				conn.Close()
-			}
+	for {
+		// Make a connection using the credentials.
+		ctx, cancel := context.WithTimeout(context.Background(), defaultConnTimeout)
+		conn, err := grpc.DialContext(ctx, address, grpc.WithTransportCredentials(clientTLSCreds))
```

Comment:
Do we need a new client inside this loop?  I expected this would not be necessary.  Maybe set keepalive so that the server closes the conn quickly?

https://godoc.org/google.golang.org/grpc/keepalive#ServerParameters

Reply:
The problem is that the custom verification, which will log the client requests on server side, only happens per-connection. If I create the client once and make multiple requests inside the loop, only one request log will be shown on the server side, and we are not able to verify the reloading behaviors...

Reply:
Even if it's per-connection, setting keepalive max connection age to something like 500ms would cause a reconnect, triggering the re-handshake & logs.

Reply:
Ah I see. Just to confirm, after the server cuts off the connection, the re-handshake wouldn't happen automatically, right(which means the client side would need to re-create the client object and restarts the connection after a connection failure)?
I might did something wrong, but from my observation, we might need to manually restart the handshake. Is this the expected behavior? If it is, we still have to create multiple client objects inside the loop...

Reply:
> Just to confirm, after the server cuts off the connection, the re-handshake wouldn't happen automatically, right(which means the client side would need to re-create the client object and restarts the connection after a connection failure)?

If the connection is closed, the client will re-establish a new connection.  You don't need to create a new `ClientConn` for that to happen.

> I might did something wrong, but from my observation, we might need to manually restart the handshake.

A new handshake will occur for every connection when it's created.  (I.e. `ClientHandshake` will be called on the `TransportCredentials` used.)

Reply:
I see, thanks for the explanation! That's an interesting behavior. I've updated the code to reflect the keepalive change. Feel free to let me know if there is anything else that could be improved!